File: explicit-java/src/main/kotlin/explicit/token/Group.kt
package explicit.token

import explicit.api.IToken
import explicit.api.Navigator
import explicit.api.extension.getLevenshteinDistance
import explicit.evaluator.ExplicitRuleEvaluation
import explicit.parser.ExplicitRules
import java.util.*
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Token implementation which represents a group of tokens.
 */
data class Group(val tokens: List<IToken>): IToken {
    override fun toString() = "[${tokens.map(IToken::asText).joinToString(" ")}]"

    override fun evaluate(rulez: ExplicitRules, navigator: Navigator<String>): ExplicitRuleEvaluation {
        val whitelist = AtomicBoolean(false)
        for (groupToken in tokens) {
            if (groupToken !is Not) {
                whitelist.set(true)
            }
            if (groupToken.asText().equals(navigator.getCurr(), ignoreCase = true)) {
                val evaluation = groupToken !is Not
                return ExplicitRuleEvaluation(evaluation, HashMap(), listOf(navigator.getIndex() - 1))
            }
            if (groupToken is Like) {
                if (groupToken.asText().toLowerCase().getLevenshteinDistance(navigator.getCurr().toLowerCase()) <= 1) {
                    return ExplicitRuleEvaluation(true, HashMap(), listOf(navigator.getIndex() - 1))
                }
            }
        }
        return ExplicitRuleEvaluation(!whitelist.get(), HashMap(), listOf(navigator.getIndex() - 1))
    }
}

File: explicit-python/explicit_nlu/token/Group.py
import logging
from dataclasses import dataclass
from typing import List

from explicit_nlu.api import IToken, Navigator
from explicit_nlu.evaluator import ExplicitEvaluation
from explicit_nlu.parser import ExplicitRules
#
# Token implementation which represents a group of tokens.
#
from explicit_nlu.token import Not, Like


@dataclass
class Group(IToken):
    tokens: List[IToken]

    def __str__(self):
        return "[" + (" ".join(list(map(IToken.as_text, self.tokens)))) + "]"

    def __repr__(self):
        return self.__str__()

    def evaluate(self, rulez: ExplicitRules, navigator: Navigator) -> ExplicitEvaluation:
        logging.debug("evaluate group")

        whitelist = False

        for token in self.tokens:
            if token is not Not:
                whitelist = True

            if token.as_text() == navigator.get_curr():
                result = token is not Not
                return ExplicitEvaluation(result, dict(), [navigator.get_absolute_index() - 1])

            if token is Like:
                if self.levenshtein(token.as_text(), navigator.get_curr()) <= 1:
                    return ExplicitEvaluation(True, dict(), [navigator.get_absolute_index() - 1])

        return ExplicitEvaluation(not whitelist, dict(), [navigator.get_absolute_index() - 1])

    # TODO: use library
    def levenshtein(self, a: str, b: str):
        a = a.lower()
        b = b.lower()
        if not a: return len(b)
        if not b: return len(a)
        return min(self.levenshtein(a[1:], b[1:]) + (a[0] != b[0]),
                   self.levenshtein(a[1:], b) + 1,
                   self.levenshtein(a, b[1:]) + 1)
